### PR TITLE
Don't fail on missing eager-loaded association

### PR DESCRIPTION
**Pull request by seanlinsley**
This allows for nested eager loading of polymorphic associations

Resolves #8005

### CODE REVIEW

File: activerecord/lib/active_record/associations/preloader.rb [hunk 1]
```diff
@@ -107,26 +107,27 @@ def preload(records, associations, preload_scope = nil)
...
+          preloaders_for_one(association.to_sym, records, scope, options)
         else
           raise ArgumentError, "#{association.inspect} was not recognised for preload"
         end
       end
 
-      def preloaders_for_hash(association, records, scope)
+      def preloaders_for_hash(association, records, scope, options = {})
         association.flat_map { |parent, child|
           loaders = preloaders_for_one parent, records, scope
```
Comment (Empact) [inline activerecord/lib/active_record/associations/preloader.rb]: I ported this code over to [activerecord_lax_includes](https://github.com/brigade/active-record-lax-includes) gem and found an issue with this line:

https://github.com/brigade/active-record-lax-includes/commit/ad07ec0bfd4617f9804222a79bcd0ed124a44eb4

File: activerecord/lib/active_record/associations/preloader.rb [hunk 2]
```diff
@@ -107,26 +107,27 @@ def preload(records, associations, preload_scope = nil)
...
+          polymorphic = options[:polymorphic] || loaders.any?{ |l| l.reflection.polymorphic? }
```
Comment (Empact) [inline activerecord/lib/active_record/associations/preloader.rb]: This line doesn't work when `l` is a `NullPreloader` - perhaps `NullPreloader` should expose the `owners` and `reflection`, as `AlreadyLoaded` does - after all they are part of the public interface of a preloader.

**Comment by mikecmpbll** [comment]
just going to throw this in to the ring; i've hacked around this in the following way:

```ruby
class Alert
  belongs_to :alertable, polymorphic: true

  belongs_to :incident, ->{ joins(:alerts).where(alerts: { id: Alert.where(alertable_type: "Incident") }) }, foreign_key: :alertable_id
  belongs_to :action, ->{ joins(:alerts).where(alerts: { id: Alert.where(alertable_type: "Action") }) }, foreign_key: :alertable_id
end
```

with these additional belongs_to associations for each alertable type, i can eager load them, and their specific nested associations like such:

```ruby
Alert.includes(incident: :categories, action: :agencies)
```

i've only just written this and haven't properly analysed it yet but i _think_ it checks out. the advantage of this approach is you can customise your eager loading per type.

**Comment by edwardmp** [comment]
Can this please be merged in?

**Comment by seanlinsley** [comment]
This is nearly a two-year-old PR at this point. I can try rebasing it, but who knows what's changed on master since this was proposed.

@sgrif is it worth the time to do that, or is this not something that you'd merge?

At the very least we should integrate https://github.com/rails/rails/pull/17479#issuecomment-214327967 into the docs, and add a reference to it in the exception that's raised.

**Comment by xhoy** [comment]
Plz i would love to use it!

**Comment by edwardmp** [comment]
@seanlinsley @xhoy indeed would be very nice to have as it's very annoying right now to use polymorphic associations for this sole issue

**Comment by BrahimDahmani** [comment]
any progress here?

**Comment by ianforsyth** [comment]
+1

**Comment by sshreibati** [comment]
+1, would like to use this

**Comment by eskrynnikov** [comment]
+1

**Comment by seanlinsley** [comment]
I understand people's frustration, but please don't +1 a ticket or PR that it's already clear is valuable / important.

As soon as I receive guidance from the Rails team, I'm willing to get this in working order again. But most of them are volunteers, so we can't expect them to dedicate time to this.

For now, @mikecmpbll's solution works fine.

**Comment by sondnm** [comment]
A simpler implementation of @mikecmpbll's solution (https://github.com/rails/rails/pull/17479#issuecomment-214327967) in Rails 4:
```ruby
class Alert
  belongs_to :alertable, polymorphic: true

  belongs_to :incident, foreign_type: 'Incident', foreign_key: :alertable_id
  belongs_to :action, foreign_type: 'Action',  foreign_key: :alertable_id
end
```

**Comment by kbrock** [comment]
preloading polymorphic relations is tricky.

```
Post.has_many :commenters, :polymorphic => true # Person or Company
Person.has_one :image
#Company does not has_one :image

Post.first.commenters.includes(:image)
```

@matthewd we had a similar issue w/ ar_virtual. I was able to properly raise errors for invalid associations, but had to revert the change. People wanted to preload associations that were only valid for some of the associated classes. [[ref]](https://github.com/ManageIQ/manageiq/pull/16579)

**Comment by rafaelfranca** [comment]
Fixed in https://github.com/rails/rails/pull/32655

**Pull request closed without merging**